src/__swaps__/utils/decimalFormatter.ts

```diff
@@ -0,0 +1,90 @@
+import {
+  ceilWorklet,
+  divWorklet,
+  floorWorklet,
+  mulWorklet,
+  roundWorklet,
+  toFixedWorklet,
+  orderOfMagnitudeWorklet,
+  significantDecimalsWorklet,
+} from '../safe-math/SafeMath';
+
+const MAXIMUM_SIGNIFICANT_DECIMALS = 6;
+const STABLECOIN_MINIMUM_SIGNIFICANT_DECIMALS = 2;
+
+export function valueBasedDecimalFormatter({
+  amount,
+  isStablecoin,
+  nativePrice,
+  niceIncrementMinimumDecimals,
+  precisionAdjustment,
+  roundingMode,
+  stripSeparators = true,
+}: {
+  amount: number | string;
+  isStablecoin?: boolean;
+  nativePrice: number;
+  niceIncrementMinimumDecimals?: number;
+  precisionAdjustment?: number;
+  roundingMode?: 'up' | 'down' | 'none';
+  stripSeparators?: boolean;
+}): string {
+  'worklet';
+
+  function calculateDecimalPlaces(): {
+    minimumDecimalPlaces: number;
+    maximumDecimalPlaces: number;
+  } {
+    const orderOfMagnitude = orderOfMagnitudeWorklet(amount);
+    let minDecimalsForOneCent = nativePrice ? Math.round(Math.max(0, Math.log10(nativePrice / 0.01))) : MAXIMUM_SIGNIFICANT_DECIMALS;
+
+    const significantDecimals = significantDecimalsWorklet(amount);
+
+    let minimumDecimalPlaces = 0;
+    let maximumDecimalPlaces = MAXIMUM_SIGNIFICANT_DECIMALS;
+    const stablecoinMin = isStablecoin ? STABLECOIN_MINIMUM_SIGNIFICANT_DECIMALS : 0;
+
+    const minBasedOnOrderOfMag = orderOfMagnitude > 2 ? 0 : 2;
+    if (orderOfMagnitude < 1) {
+      minimumDecimalPlaces = Math.max(stablecoinMin, significantDecimals);
+      maximumDecimalPlaces = Math.max(minBasedOnOrderOfMag, minDecimalsForOneCent, significantDecimals + 1);
+    } else {
+      minimumDecimalPlaces = stablecoinMin;
+      maximumDecimalPlaces = Math.max(minBasedOnOrderOfMag, minDecimalsForOneCent - orderOfMagnitude);
+    }
+
+    return {
+      minimumDecimalPlaces,
+      maximumDecimalPlaces: Math.max(
```

Comment:
This just ensures that after all of our ideal calculations, we ensure that:
1. incorporate any `precisionAdjustment` here if wanted
2. ensure that if there's a value passed in from the `niceIncrementFormatter` that we at least are +1 decimal above that
3. the max should always be greater than the min decimals calculated